design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. If the columns contain `ROWKEY`, the
+value will be verified against the key schema but will not populate the corresponding field in the 
+value. Populating columns with `ROWTIME` will not be supported.
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Key Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT INTO`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. To all for all levels of flexibility, `INSERT INTO` should allow
+for all cases that kafka supports. Therefore, `ROWKEY` will always be available in the list of
+column names as the first column. The following examples would all be possible: 
+
+```sql
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", null, 123);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", "k", 456);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES (null, "k", 789);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", "x", 1011); 
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", null, null); 
+```
+
+If `ROWKEY` is _not_ specified but the key field is specified (as defined by the corresponding 
```

Comment:
Thinking about this some more I think we should say: 

1. For `TABLE`: 
  * the key is set to that of the column specified in the `KEY` of the CT. 
  * `ROWKEY` is not valid in `INSERT INTO` and will throw an exception
2. For `STREAM`: 
  * Since we don't use `KEY` in CSAS/CS currently we should not introduce it now
  * User can specify `ROWKEY` in `INSERT INTO`, optionally
  * If user doesn't speicfy `ROWKEY` in `INSERT INTO` then null key is written

Reply:
Based on the `syntax-reference`, we do support `KEY` in the `WITH` clause for streams. Given that, to be more consistent with `TABLE` semantics, we could just use that key if present otherwise set it to null. Thoughts?

My only concern regarding keys (cc @big-andy-coates) is that I believe we're removing the requirement for setting the key in `CT` statements (if it is absent, we will just use whatever the kafka key is). If we make these types of changes in the future, then this implicit keying might need to be changed.

Reply:
As per comment above, we need to support inserting values into a stream _and_ table where no KEY field was specified in the WITH clause of the CREATE statement.

Just setting it to `null` is not an option IMHO, as its too restrictive. `WITH(KEY)` is an optimisation / compiler hint, (or soon will be), and so this new syntax should support the user supplying an explicit key.

Also, any syntax we come up with for supplying said key would ideally be extendable to structured keys.

I think the syntax needs to support the user being explicit about they key of the row. Thinking back to the query semantics discussions we've been having I think we agreed on using the `KEY` keyword in CT/CS statements to define which columns make up the key and which the value.  We should probably look to do something similar for INSERT VALUES.

```sql
INSERT INTO foo (col0 KEY, col1 KEY, col2, col3) VALUES (a, b, c, d);
```

Though we'd need to think how this would work when the schema is not supplied.

Where the source we're inserting into has a key field defined, (and remember the source may not have been created by a CT/CS statement - it may have been created by a CSAS/CTAS statement), then we can look to not require the key, or at the very least assert that the two columns match.

Reply:
Same as above (pardon my flip-flopping, updated version handles things a little cleaner). 

If possible, though, I'd prefer we didn't introduce `KEY` in this until we have structured keys implemented (we can stage it so that structured key support doesn't come to `INSERT INTO ... VALUES` until some time afterward it is supported generally).